dtm0/fop.c

```diff
@@ -110,46 +115,91 @@ int m0_dtm0_fop_init(void)
 			 .opcode    = M0_DTM0_REQ_OPCODE,
 			 .xt        = dtm0_req_fop_xc,
 			 .rpc_flags = M0_RPC_MUTABO_REQ,
-			 .fom_ops   = &dtm0_req_fop_fom_type_ops,
+			 .fom_ops   = &dtm0_req_fom_type_ops,
 			 .sm        = &m0_generic_conf,
 			 .svc_type  = &dtm0_service_type);
 	M0_FOP_TYPE_INIT(&dtm0_rep_fop_fopt,
 			 .name      = "DTM0 reply",
 			 .opcode    = M0_DTM0_REP_OPCODE,
 			 .xt        = dtm0_rep_fop_xc,
 			 .rpc_flags = M0_RPC_ITEM_TYPE_REPLY,
-			 .fom_ops   = &dtm0_req_fop_fom_type_ops);
+			 .fom_ops   = &dtm0_req_fom_type_ops);
 	return 0;
 }
 
+static bool dtm0_service_started(struct m0_fop  *fop,
+				struct m0_reqh *reqh)
+{
+	struct m0_reqh_service            *svc;
+	const struct m0_reqh_service_type *stype;
+
+	stype = fop->f_type->ft_fom_type.ft_rstype;
+	M0_ASSERT(stype != NULL);
+
+	svc = m0_reqh_service_find(stype, reqh);
+	M0_ASSERT(svc != NULL);
+
+	return m0_reqh_service_state_get(svc) == M0_RST_STARTED;
+}
 /*
-  Allocates and initialises a fom.
+  Allocates a fop.
  */
-static int dtm0_req_fop_fom_create(struct m0_fop *fop,
-				   struct m0_fom **out, struct m0_reqh *reqh)
-{
-        struct m0_fom           *fom;
-	struct m0_fop           *rfop;
 
-	M0_PRE(fop != NULL);
-        M0_PRE(out != NULL);
-	M0_PRE(M0_IN(m0_fop_opcode(fop), (M0_DTM0_REQ_OPCODE)));
+int m0_dtm0_fop_create(struct m0_rpc_session	 *session,
+		       enum m0_dtm0s_msg	  opmsg,
+		       struct m0_dtm0_tx_desc	 *txr,
+		       struct m0_fop		**out)
+{
+	struct dtm0_req_fop  *op;
+	struct m0_fop	     *fop;
+	int                rc = 0;
 
-        M0_ALLOC_PTR(fom);
-        if (fom == NULL)
-                return -ENOMEM;
+	*out = NULL;
 
-	rfop = m0_fop_reply_alloc(fop, &dtm0_rep_fop_fopt);
+	fop = m0_fop_alloc_at(session, &dtm0_req_fop_fopt);
+	if (fop == NULL)
+		return M0_ERR(-ENOMEM);
+	op = m0_fop_data(fop);
+	op->dtr_msg = opmsg;
+	op->dtr_txr = *txr;
+	*out = fop;
+	return rc;
+}
+M0_EXPORTED(m0_dtm0_fop_create);
 
-	if (rfop == NULL) {
+/*
+  Allocates a fom.
+ */
+static int dtm0_fom_create(struct m0_fop *fop,
+			  struct m0_fom **out, struct m0_reqh *reqh)
+{
+	struct dtm0_fom    *fom;
+	struct m0_fom     *fom0;
+	struct m0_fop     *repfop;
+	struct dtm0_rep_fop *reply;
+
+	if (!dtm0_service_started(fop, reqh))
+		return M0_ERR(-EAGAIN);
```

Comment:
Do we really need this check?
I feel that we never hit -EAGAIN case here. Also, I think that similar check is somewhere during processing on the fom checking that service instance is proper.

Reply:
I think this may not be needed .. will remove it.